### PR TITLE
Capsula js transport providers

### PR DESCRIPTION
**Pull request by taras-shedenko**

### CODE REVIEW

File: src/HTTP/AxiosDispatcher.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import axios from 'axios';
+
+import { Dispatcher } from '../api';
+
+export class AxiosDispatcher extends Dispatcher {
+  dispatch<T, R>(api: string, request: T): Promise<R> {
+    return axios.post(this.baseUrl + api, request)
+      .then(response => response.data)
+      .catch(error => Promise.reject(error.response.data));
```
Comment (dmitriyslaym) [inline src/HTTP/AxiosDispatcher.ts]: are you sure that the structure for error object will be always the same?

File: src/WebSocket/WebSocketDispatcher.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
+/*
+*  Send and receive requests serially
+*  next reuqest is only sent when the previous one got its response
```
Comment (dmitriyslaym) [inline src/WebSocket/WebSocketDispatcher.ts]: reuqest typo

File: src/WebSocket/WebSocketDispatcher.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+*/
+
+import { Dispatcher } from '../api';
+
+const errorMessages = {
```
Comment (dmitriyslaym) [inline src/WebSocket/WebSocketDispatcher.ts:8]: better to create a separate file for errors

File: src/WebSocket/WebSocketDispatcher.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+  notConnected: 'Not connected!',
+  resourceIsBusy: 'Resource is busy!',
```
Comment (dmitriyslaym) [inline src/WebSocket/WebSocketDispatcher.ts:10]: do we need exclamations?

File: src/WebSocket/WebSocketDispatcher.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  unknownState: 'Error due to unknown state',
+};
+
+export class WebSocketDispatcher extends Dispatcher {
+  private webSocket?: WebSocket;
+  private responseData: any;
+
+  open(): Promise<null> {
```
Comment (dmitriyslaym) [inline src/WebSocket/WebSocketDispatcher.ts]: why you put null? Don't put anything in resolve and the returning type should be void

File: src/WebSocket/WebSocketDispatcher.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    return new Promise((resolve, reject) => {
+      this.webSocket = new WebSocket(this.baseUrl);
+      this.webSocket.onopen = () => resolve(null);
+      this.webSocket.onerror = error => reject(error);
+    });
+  }
+
+  dispatchInt<T, R> (api: string, request: T): Promise<R> {
```
Comment (dmitriyslaym) [inline src/WebSocket/WebSocketDispatcher.ts:31]: suggest to mark inner methods as _
**_dispatch**

File: src/WebSocket/WebSocketDispatcher.ts (hunk 5 above)
Comment (dmitriyslaym) [reply src/WebSocket/WebSocketDispatcher.ts]: is open an inner method? in this case it's **_open**

File: src/WebSocket/WebSocketDispatcher.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    });
+  }
+
+  dispatch<T, R> (api: string, request: T): Promise<R> {
+
+    switch (this.getState()) {
+      case 'OPEN': return this.dispatchInt(api, request);
+      case 'NONE':
+      case 'CLOSED':
+        return this.open().then(() =>
```
Comment (dmitriyslaym) [inline src/WebSocket/WebSocketDispatcher.ts:77]: what if there is an error while calling this.open()? Do you handle it anywhere?

File: src/WebSocket/WebSocketDispatcher.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+          this.dispatchInt<T, R>(api, request)
+        );
+      default: return Promise.reject(errorMessages.unknownState);
```
Comment (dmitriyslaym) [inline src/WebSocket/WebSocketDispatcher.ts:80]: do you want to wrap it in new Error()?

File: src/WebSocket/WebSocketDispatcher.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    };
+  }
+
+  finalize(): Promise<null> {
```
Comment (dmitriyslaym) [inline src/WebSocket/WebSocketDispatcher.ts]: the same for **null** response

File: src/WebSocket/WebSocketDispatcher.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    return new Promise((resolve, reject) => {
+      webSocket.onclose = () => resolve(null);
+      webSocket.onerror = () => reject();
+      webSocket.close();
+    });
+  }
+
+  getState(): string {
```
Comment (dmitriyslaym) [inline src/WebSocket/WebSocketDispatcher.ts:99]: is it an inner method?

File: src/api/Dispatcher.ts [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+export abstract class Dispatcher {
+  // Base URL
+  protected baseUrl: string;
+
+  // Base Constructor
+  constructor(baseUrl: string) {
+    this.baseUrl = baseUrl;
+  };
+
+  // Interface for the Dispatcher
```
Comment (dmitriyslaym) [inline src/api/Dispatcher.ts:12]: please provide more description here

File: tests/sanity.test.ts [hunk 12]
```diff
@@ -0,0 +1,15 @@
+import { AxiosDispatcher } from 'HTTP/AxiosDispatcher';
```
Comment (dmitriyslaym) [inline tests/sanity.test.ts:1]: I think that there are many things to be tested here. Or is it not a part of the task?

**Review by dmitriyslaym — COMMENTED** [review]
Please take a look at all the comments.

File: src/HTTP/AxiosDispatcher.ts (hunk 1 above)
Comment (taras-shedenko) [reply src/HTTP/AxiosDispatcher.ts]: That's right: response is optional in AxosError

File: src/WebSocket/WebSocketDispatcher.ts (hunk 2 above)
Comment (taras-shedenko) [reply src/WebSocket/WebSocketDispatcher.ts]: fixed

File: src/WebSocket/WebSocketDispatcher.ts (hunk 3 above)
Comment (taras-shedenko) [reply src/WebSocket/WebSocketDispatcher.ts:8]: There is no need to export/import those values

File: src/WebSocket/WebSocketDispatcher.ts (hunk 4 above)
Comment (taras-shedenko) [reply src/WebSocket/WebSocketDispatcher.ts:10]: Why not? It is just to emphasize the severity of the message!

File: src/WebSocket/WebSocketDispatcher.ts (hunk 5 above)
Comment (taras-shedenko) [reply src/WebSocket/WebSocketDispatcher.ts]: open method could be used externally
type null means no value, void is just the same

File: src/WebSocket/WebSocketDispatcher.ts (hunk 8 above)
Comment (taras-shedenko) [reply src/WebSocket/WebSocketDispatcher.ts:80]: Why?

File: src/WebSocket/WebSocketDispatcher.ts (hunk 9 above)
Comment (taras-shedenko) [reply src/WebSocket/WebSocketDispatcher.ts]: The same

File: src/WebSocket/WebSocketDispatcher.ts (hunk 10 above)
Comment (taras-shedenko) [reply src/WebSocket/WebSocketDispatcher.ts:99]: Let keep available for external usage

File: tests/sanity.test.ts (hunk 12 above)
Comment (taras-shedenko) [reply tests/sanity.test.ts:1]: It's just an network protocols wrapper, main tests supposed to be in services

File: src/WebSocket/WebSocketDispatcher.ts (hunk 3 above)
Comment (dmitriyslaym) [reply src/WebSocket/WebSocketDispatcher.ts:8]: There will be a need for it if we want to cover the functionality by tests

File: src/WebSocket/WebSocketDispatcher.ts (hunk 5 above)
Comment (dmitriyslaym) [reply src/WebSocket/WebSocketDispatcher.ts]: I remember Idan didn't like null in Promise, confirm with him please.
If a method is not described in a public interface, we don't want it to be used externally, so we mark it with _

File: tests/sanity.test.ts (hunk 12 above)
Comment (dmitriyslaym) [reply tests/sanity.test.ts:1]: I had many basic tests in scalecube, when I was working on Transport and its providers. Suggest to wait Idans confirmation that this amount of tests is fine, cause I feel it's not.

File: src/WebSocket/WebSocketDispatcher.ts (hunk 5 above)
Comment (taras-shedenko) [reply src/WebSocket/WebSocketDispatcher.ts]: void - means no type, mostly used for functions without any return statement
null - for cases when you want to specify "no-value"
Promise<void>  - seems incorect, Promise<null> - much more meaningfull

Comment (taras-shedenko) [reply src/WebSocket/WebSocketDispatcher.ts]: You were right, Dmitriy. Promise.resolve() is of type Promise<void>, not Promise<null>
Need to re-factor a little.

**Pull request merged**